7a_temp_coop_munge/src/data_parsers/parse_Iowa_DNR_LimnoProfiles_2000_2020.R

```diff
@@ -0,0 +1,507 @@
+  tc_out <- tryCatch(
+    { read_iowa_xls_tbl(x, keep_cols, data_start = 3) }, # current thinking is that all true xls start on row 3
+    error = function(cond) {
+      if(htm_verbose) {print(paste(basename(x), ' is in HTM format', sep = ''))}
+      read_iowa_htm_tbl(x, keep_cols, htm_datetime, htm_data_start)
+    }
+  )
+
+  # add LakeID and date columns from file name if necessary
+  tc_out$lakeid <- extract_lakeid(x, start = subset_lakeid[1], stop = subset_lakeid[2])
+
+  return(tc_out)
+}
+
+#' Parse Iowas data from 2017-2020
+#'
+#' @param x chr, full file path
+#' @param keep_cols chr, vector of column names to keep.
+#' @param lakeid_col num, Column index for the location of the Lake ID field. The column that is used to ID the Iowa lake moves around and changes names from file to file.
+#' @param temp_as_f logical, should temp values be converted from F to C?
+#' @param use_readr logical, not all files are read well by `readr::read_csv`. If `use_readr = F` then `read.csv` is used.
+#' Column name spelling should match. Case does not matter because
+#' all values to be changed `tolower` within the function.
+parse_2017_2020_data <- function(x, keep_cols, lakeid_col, temp_as_f = F, use_readr = T){
+
+  if(grepl('*xlsx', x)) {
+    dat <- readxl::read_xlsx(x)
+  } else if (use_readr){
+    dat <- readr::read_csv(x)
+  } else {
+    dat <- read.csv(x)
+  }
+
+  # select key columns, normalize
+  dat <- dat %>% dplyr::select(all_of(keep_cols))
+  names(dat) <- tolower(names(dat))
+
+  # validate fields
+  if(!any(names(dat) %in% c('lake', 'site'))) {
+    # this is specifically to accommodate one file: "2019_R3_Profiles.csv"
+    # 2017 data is generally called 'site' and 2018-2020 is 'lake' for the id field
+    names(dat)[lakeid_col] <- 'lake'
+    idx_date <- grepl('*date*', names(dat)) %>% which(. %in% "TRUE")
+    names(dat)[idx_date] <- 'date'
+  }
+
+  # add numeric lakeid - this comes in a variety of flavors between 2017-2020,
+  # just extracting the numbers
+  dat$lakeid <- sapply(dat[ , lakeid_col], function(x) {
+    as.numeric(gsub(".*?([0-9]+).*", "\\1", x))
+  })
+
+  # handle date
+  if(any(class(dat$date) == 'character')) {
+    dat$date <- lubridate::mdy(dat$date) %>% as.Date
+  }
+
+  # check for datetimes and AM/PM designation
+  if(any(sapply(dat$time, function(x) class(x) == "POSIXct"))) {
+
+    dat$time <- strftime(dat$time, format = "%H:%M") %>% as.character()
+
+  } else if(any(sapply(dat$time, function(x) nchar(x) > 9))) {
+
+    dat$time <- ifelse(nchar(dat$time) == 11,
+                       substr(dat$time, 1, 5),
+                       substr(dat$time, 1, 4))
+
+  } else {
+
+    # 2018-2020 data has a little extra whitespace and often includes the seconds
+    dat$time <- format(dat$time, '%H:%M') %>%
+      as.character(.) %>%
+      substr(., 1, 5)
+
+  }
+
+  # convert temp if necessary
+  if(temp_as_f) {
+    idx <- grepl('*temp*', names(dat)) %>% which(. %in% "TRUE")
```

Comment:
As I'm digging into these files, I see why all of the funky partial matching is necessary 😱

Reply:
Should I keep it or shift to `case_when`? In general, I try to use `base` solutions if aren't too convoluted

Reply:
Also, I'll add some more doc here

Reply:
Complicated question ("Should I keep it or shift")! 

I generally try to stay in dplyr-land for these munging codes since it is easier to read and understand in general. But I do also agree that we should stay in base when there is an adequate solution and no good reason to bring in a _new_ package to provide a single function. I think our general (unwritten ☹️ ) coding styles for the branch would suggest this munging code would be in dplyr, plots are generally ggplot2, but all kinds of space to deviate from that pattern as needed. 

But this is probably something others should weigh in on. At the project lead level, @lindsayplatt do you have opinions? I also think there is room to suggest a future pattern without asking for a re-write of this particular code. This is related to my [overarching comment above](https://github.com/USGS-R/lake-temperature-model-prep/pull/284#pullrequestreview-859450328) that was
> A general question, for the partial matching and indexing and reassigning with base R, is there a way to do some of these operations with case_when and tidyselect matching like contains and starts_with to keep things dplyr'd?

Reply:
I have been staying in tidyverse-land lately with projects due to its readability, but I think its true that there is a critical point when you are starting a project where you need to think about the benefits of loading in a new package(s). For this project, we already use `dplyr` and other `tidyverse` functions in a number of places, so I think it makes sense to lean into that practice. Times when we might want to deviate are when the dplyr commands are significantly slower (which can happen with some bigger data). In those cases, we may turn to base R or to something like `data.table`.

Reply:
I have definitely been working to avoid new packages. How about we split the difference on this one and I start using `case_when` into the future. I suggest that for this specific case because there are so many edge cases in this dataset (hence all the partial matching and `ifelses`) within this dataset. It would take a bit of time to resolve it. To compensate for any clarity issues I can add more into the comments.

Reply:
I think extra comments in lieu of converting to `case_when()` would be a good compromise in this instance. Maybe also make a note about this very decision somewhere in the code - choosing to forgo a conversion to `case_when()` in order to keep moving forward (could even link to this discussion by clicking on the time next to the username to get a link specifically to these comments).

Reply:
I'm good with this too and hope my original comment was worded such that it was clear I thought this change would be totally optional here, but something to consider in the future. Glad we had this discussion!

Reply:
Sidebar chat with Lindsay. We are going to leave this as is within this parser, but I will use `case_when` in future parsers